neuralpp/symbolic/z3_expression.py

```diff
@@ -498,7 +512,115 @@ def __and__(self, other: Any) -> Context:
                 other = Z3Expression.convert(other)
             # Always treat `other` as a literal. Will raise if it cannot be converted to a boolean in Z3.
             new_solver = z3_add_solver_and_literal(self._solver, other.z3_object)
-            new_dict = self._dict | _extract_key_value_from_assertions([other.z3_object])
+            new_dict = self._dict | _extract_key_to_value_from_assertions([other.z3_object])
         return Z3SolverExpression.make(new_solver, new_dict)
 
     __rand__ = __and__
+
+    @cached_property
+    def variable_replacement_dict(self):
+        """
+        Since equality is transitive, a set of equalities would create a set of elements who are equal to each other.
+        We refer to such a set as an "equivalence class". If we have a total order "<=" on the elements of the
+        equivalence class, we can compute a minimum of that class. Note this total order "<=" has nothing to do with
+        the equivalence relation.
+        The elements in an equivalence class can be variables and at most one value (if two distinct value a and b
+        exist in an equivalence class, then the context is unsatisfiable). A value is always smaller than all variables.
+
+        A variable_replacement_dict is a Variable -> (Variable | Value) dictionary that maps a variable in the context
+        to the minimum of its equivalence class. E.g., suppose the total order of variables is the alphabetical order,
+        and the context has the following assertions: { x == y, y > 1, z == y, a == 3, b == a },
+        then variable_replacement_dict would be { x: x, y: x, z: x, a: 3, b: 3}.
+        """
+        return {_z3_object_to_expression(key): _z3_object_to_expression(value)
+                for key, value in self.z3_variable_replacement_dict.items()}
+
+    @cached_property
+    def z3_variable_replacement_dict(self):
+        """
+        `variable_replacement_dict`, except in z3's native representation
+        """
+        result = {}
+        equivalence_classes = _extract_equivalence_classes_from_assertions(self.assertions)
+        for equivalence_class in equivalence_classes:
+            for element in equivalence_class:
+                result[element] = equivalence_class.minimum
+        return result
+
+
+class EquivalenceClass(frozenset):
+    def __eq__(self, other):  # either same instance or not equal, no need to use more expensive `==`.
+        return self is other
+
+    __hash__ = frozenset.__hash__  # otherwise it's set to None by default
+
+    @cached_property
+    def minimum(self) -> Any:
+        """ Assumes len(self) > 0. """
+        min_value = None
+        for element in self:
+            min_value = element if min_value is None else min(element, min_value)
+        return min_value
```

Comment:
Looks like we can simply return `min(self)` here?

Reply:
Yes. The old design here contains a bug so I rewrote this part. In fixed version I use `min()`.